flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/backpressure/BackPressureRequestCoordinatorTest.java

```diff
@@ -0,0 +1,334 @@
+
+		List<CompletableFuture<BackPressureStats>> requestFutures = new ArrayList<>();
+
+		// trigger request
+		requestFutures.add(coordinator.triggerBackPressureRequest(vertices));
+		requestFutures.add(coordinator.triggerBackPressureRequest(vertices));
+
+		for (CompletableFuture<BackPressureStats> future : requestFutures) {
+			assertFalse(future.isDone());
+		}
+
+		// shut down
+		coordinator.shutDown();
+
+		// verify all completed
+		for (CompletableFuture<BackPressureStats> future : requestFutures) {
+			assertTrue(future.isCompletedExceptionally());
+		}
+
+		// verify new trigger returns failed future
+		CompletableFuture<BackPressureStats> future = coordinator.triggerBackPressureRequest(vertices);
+		assertTrue(future.isCompletedExceptionally());
+	}
+
+	private ExecutionVertex[] createExecutionVertices(
+			ExecutionState state,
+			CompletionType completionType) throws Exception {
+		return createExecutionVertices(state, completionType, requestTimeout);
+	}
+
+	private ExecutionVertex[] createExecutionVertices(
+			ExecutionState state,
+			CompletionType completionType,
+			long requestTimeout) throws Exception {
+		return new ExecutionVertex[] {
+			createExecutionVertex(0, ExecutionState.RUNNING, CompletionType.SUCCESSFULLY, requestTimeout),
+			createExecutionVertex(1, state, completionType, requestTimeout),
+			createExecutionVertex(2, ExecutionState.RUNNING, CompletionType.SUCCESSFULLY, requestTimeout)
+		};
+	}
+
+	private ExecutionVertex createExecutionVertex(
+			int subTaskIndex,
+			ExecutionState state,
+			CompletionType completionType,
+			long requestTimeout) throws Exception {
+		return new TestingExecutionVertex(
+			BackPressureTrackerTestUtils.createExecutionJobVertex(),
+			subTaskIndex,
+			new IntermediateResult[0],
+			Time.seconds(10),
+			1L,
+			System.currentTimeMillis(),
+			JobManagerOptions.MAX_ATTEMPTS_HISTORY_SIZE.defaultValue(),
+			state,
+			completionType,
+			requestTimeout);
+	}
+
+	/**
+	 * Completion type of the request future.
+	 */
+	private enum CompletionType {
+		SUCCESSFULLY,
+		EXCEPTIONALLY,
+		TIMEOUT,
+		NEVER_COMPLETE
+	}
+
+	/**
+	 * A testing {@link ExecutionVertex} implementation used to wrap {@link TestingExecution}.
+	 */
+	private static class TestingExecutionVertex extends ExecutionVertex {
+
+		private final Execution execution;
+
+		public TestingExecutionVertex(
+				ExecutionJobVertex jobVertex,
+				int subTaskIndex,
+				IntermediateResult[] producedDataSets,
+				Time timeout,
+				long initialGlobalModVersion,
+				long createTimestamp,
+				int maxPriorExecutionHistoryLength,
+				ExecutionState state,
+				CompletionType completionType,
+				long requestTimeout) {
+
+			super(
+				jobVertex,
+				subTaskIndex,
+				producedDataSets,
+				timeout,
+				initialGlobalModVersion,
+				createTimestamp,
+				maxPriorExecutionHistoryLength);
+			execution = new TestingExecution(
```

Comment:
From this logic, the real created `Execution` in `TestingExecutionVertex` is different from `TestingExecution`.
Then I found that it would be simpler for tests purpose if we introduce `ExecutionVertex#requestBackPressure` in the core code path. If so, we can override the method of `requestBackPressure` in this `ExecutionVertex` level to remove the class `TestingExecution` completely, and call `getCurrentExecutionAttempt().transitionState(state)` here. 

From `BackPressureRequestCoordinator#triggerBackPressureRequest`, it also gets the array of `ExecutionVertex` as argument now and it is more naturally to call request on `ExecutionVertex` directly.

But considering the core code path almost ready, so I can accept the current now if it would touch more changes.

Reply:
I agree with you that we can keep the current implementation and avoid modification of the core code.